libtsuba/include/tsuba/Cache.h

```diff
@@ -0,0 +1,182 @@
+#ifndef KATANA_LIBTSUBA_TSUBA_CACHE_H_
+#define KATANA_LIBTSUBA_TSUBA_CACHE_H_
+
+// This is single threaded only, it is not intended to store large objects,
+// but rather metadata (e.g., a shared_ptr to a property column).
+
+// The problem witchel had implementing a multi-threaded version using
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/ArrowInterchange.h"
+#include "katana/ConcurrentMap.h"
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+struct KATANA_EXPORT KeyBase {
+  enum class NodeEdge { kNode, kEdge };
```

Comment:
I think this issue is covered more completely below, but I need three things from the key, one of which is whether it is a node property or an edge property.

Reply:
I'm still not sure why this is here as the Cache class doesn't really care what the key is as long as it is hashable. `KeyBase` sounds like something that goes with the code making the Cache not here where we define the Cache class for everyone.

All your Key type needs to do is implement the Hash protocol: https://en.cppreference.com/w/cpp/utility/hash

Reply:
It was still there because I thought it needed to be there, but you are right it doesn't.

I have removed it, but a consequence is that I have to "inject" its hash function into namespace `std`.  That has two disadvantages to my eyes: (1) the fields of the key have to be public so the hash function can read them and (2) it feels unencapsulated to me.  I wish the hash function could be defined as part of the object itself.

Neither of these issues is large, but I would be interested to hear your perspective.

Reply:
> I have removed it, but a consequence is that I have to "inject" its hash function into namespace std. That has two disadvantages to my eyes: (1) the fields of the key have to be public so the hash function can read them

You can either make your std::hash specialization a friend of your key class or just have it call a function you want on the key class itself that then returns the hash.

> and (2) it feels unencapsulated to me. I wish the hash function could be defined as part of the object itself.

You can make the has function part of the object itself. You just need to tell the standard library what that function is called and you do that by extending `std::hash` as described.

Reply:
Ah hah. I like the idea of making it a function called by `std::hash`.  That wasn't the pattern used on the cppreference website, though I should have thought of it.